SalesforceCommerceCloud/pwa-kit
# Fix server error handling for unrecoverable errors

# Description



_Previous attempt: https://github.com/SalesforceCommerceCloud/pwa-kit/pull/316_

Currently there is an issue where if an error is thrown from `AppConfig`, the error will not be caught and cause the app to hang.

To reproduce this issue, add `throw new Error('error')` to your AppConfig render method.

## Definitions

Before we discuss the solution, we need to clarify two terms unrecoverable error V.S. recoverable error.

A **recoverable** error is the error which was thrown from rendering the React app, such as the routable component's render methods, and are caught by the AppErrorBoundary, such errors are then rendered on the error page.

A **unrecoverable** error is the error which was thrown from a component that is not wrapped by the AppErrorBoundary, such as the `AppConfig` or `Error` component. Our existing error page cannot handle such error and this PR is about handing the unrecoverable errors.

## Existing error handling behavior

1. render the react application server side
2. if error happens (both recoverable and unrecoverable)
3. re-render the app with error page (PROBLEM: if error is unrecoverable, server hangs.)

Problem with the approach is that, there is no error handling at step no.3 when the rendering of the error page fails.

Today, because the AppConfig component is not wrapped within the error boundary, errors thrown from AppConfig will cause the server to hang since the code doesn't handle such errors.

## New error handling behavior

1. render the react application server side
2. if error happens (both recoverable and unrecoverable)
3. re-render the app with error page <- we handle the recoverable error
4. if error happens again <- this is the unrecoverable errors
5. catch the error and call Express error middleware to handle unrecoverable error `next(error)` <- we handle the unrecoverable error

On development environment, the default Express error middleware returns the error response (including error stack) :
<img width="902" alt="Screen Shot 2022-01-17 at 6 02 50 PM" src="https://user-images.githubusercontent.com/10948652/149999005-e035a6b1-9ad0-4e16-8b0a-6a2b930be7b2.png">

On Production, it returns "Internal Server Error", see it live here: https://scaffold-pwa-react17.mobify-storefront.com/

I believe the default Express error middleware satisfy our requirement for handling the unrecoverable errors. Let me know if you think it is not the case.

And with my changes, these testing errors are not silenced causing unit tests to fail. I'm trying to to mock `ChunkExtractor` properly.

# Types of Changes




> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- Update server side rending logic to handle unrecoverable errors
- Remove code that swallows error for testing purposes, instead, added mocks for `loadable/server`
- refactor code and created a new function `renderAppHtml`

# How to Test-Drive This PR

1. test unrecoverable error handling
- `npm ci` at root
- add `throw new Error()` in `/packages/pwa/components/_app-config` render method
- `cd packages/pwa && npm start`
- go to localhost:3000 and verify the server returns an error response (unstyled)

2. test recoverable error handling
- - add `throw new Error()` in `/packages/pwa/pages/home/index.jsx` render method
- `npm start`
- verify you see the styled error page 

# Checklists




## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
